### ISSUE TITLE
buf generate: fails on symlinks without proper permissions

### ISSUE THREAD

**Issue by adambabik** [bufbuild/buf#265]
I tried to run `buf generate` with one dependency:
```
deps:
  - buf.build/beta/googleapis
```

and a broken symlink in the root project directory.

The command failed on resolving the symlink:
```
buf generate
buf: downloading buf.build/beta/googleapis:1dc4674e3cb949b388204fa2dc321be7
lstat /root/.cache: permission denied
```

Where a symlink pointed at: `/root/.cache/bazel/path/to/file`.

**Comment by johanbrandhorst** [comment]
I saw this error too, I think it may have something to do with how the grpc-gateway uses a docker container to generate the files. `buf` will look for `$CACHE_DIR` and `$XDG_CACHE_HOME` or otherwise fall back to `$HOME/.cache`, so if this error happens, it's likely something to do with the local configuration. See https://github.com/bufbuild/buf/blob/4d19835b0a74b71386740554f17de707514327eb/internal/pkg/app/app_unix.go#L70 and https://github.com/bufbuild/buf/blob/4d19835b0a74b71386740554f17de707514327eb/internal/pkg/app/appname/container.go#L82.

Another note, because running generate inside a container can mess with permissions, you may need to run `sudo chown $(whoami): -R .` after regenerating files.

**Comment by adambabik** [comment]
I found this code and tried to use these envs but with no luck (I tried also with `CACHE_DIR`):

```
➜ export XDG_CACHE_HOME=$HOME/.cache
➜ echo $XDG_CACHE_HOME 
/home/adamb/.cache
➜ buf generate     
lstat /root/.cache: permission denied
```

You are right with the docker and permissions being broken but after I fixed it, it's the same error. I can try to dig into the code of buf to find a code path that does not work if it's hard to reproduce. Like I said above, it works like a charm on macOS.

**Comment by johanbrandhorst** [comment]
I can't reproduce this error locally anymore, but it would be super helpful if you could keep your environment broken and help us debug what's happening here. I'd add some print statements to the code to see why it's choosing that path and rebuild, if you want I can provide the binaries for you to try?

**Comment by adambabik** [comment]
I talked with @johanbrandhorst on Buf's Slack. We found that the root cause was a broken symlink pointing at a directory to which the current user had no permissions. I updated the description.

**Comment by johanbrandhorst** [comment]
In the next release we're making the error message around this a bit clearer so the user knows to look at the contents of their repository.

**Comment by bufdev** [comment]
An updated error message is included in v0.37.1.
